velox/common/caching/SsdFile.h

```diff
@@ -0,0 +1,297 @@
+  // used only for updating statistics. Allows implementing coalesced
+  // read of nearby entries.
+  void read(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>> buffers,
+      int32_t numEntries);
+
+  // Increments the pin count of the region of 'offset'.
+  void pinRegion(uint64_t offset);
+
+  // Decrements the pin count of the region of 'offset'. If the pin
+  // count goes to zero and evict is due, starts the evict.
+  void unpinRegion(uint64_t offset);
+
+  // Asserts that the region of 'offset' is pinned. This is called by
+  // the pin holder. The pin count can be read without mutex.
+  void checkPinned(uint64_t offset) const {
+    VELOX_CHECK_LT(0, regionPins_[regionIndex(offset)]);
+  }
+
+  // Returns the region number corresponding to offset.
+  static int32_t regionIndex(uint64_t offset) {
+    return offset / kRegionSize;
+  }
+
+  // Updates the read count of a region.
+  void regionUsed(int32_t region, int32_t size) {
+    regionScore_[region] += size;
+  }
+
+  int32_t maxRegions() const {
+    return maxRegions_;
+  }
+
+  int32_t shardId() const {
+    return shardId_;
+  }
+
+  // Adds 'stats_' to 'stats'.
+  void updateStats(SsdCacheStats& stats) const;
+
+  // Resets this' to a post-construction empty state. See SsdCache::clear().
+  void clear();
+
+ private:
+  static constexpr int32_t kDecayInterval = 1000;
+
+  // Increments the pin count of the region of 'offset'. Caller must hold
+  // 'mutex_'.
+  void pinRegionLocked(uint64_t offset) {
+    ++regionPins_[regionIndex(offset)];
+  }
+
+  // Returns [start, size] of contiguous space for storing data of a
+  // number of contiguous 'pins' starting with the pin at index
+  // 'begin'.  Returns a run of 0 bytes if there is no space. The
+  // space does not necessarily cover all the pins, so multiple calls
+  // starting at the first unwritten pin may be needed.
+  std::pair<uint64_t, int32_t> getSpace(
+      const std::vector<CachePin>& pins,
+      int32_t begin);
+
+  // Removes all 'entries_' that reference data in regions described by
+  // 'regionIndices'.
+  void clearRegionEntriesLocked(const std::vector<int32_t>& regionIndices);
+
+  // Clears one or more  regions for accommodating new entries. The regions are
+  // added to 'writableRegions_'. Returns true if regions could be cleared.
+  bool growOrEvictLocked();
+
+  // Increments event count and periodically decays scores.
+  void newEventLocked();
+
+  // Verifies that 'entry' has the data at 'run'.
+  void verifyWrite(AsyncDataCacheEntry& entry, SsdRun run);
+
+  // Serializes access to all private data members.
+  std::mutex mutex_;
+
+  // Shard index within 'cache_'.
+  int32_t shardId_;
+
+  // Number of kRegionSize regions in the file.
+  int32_t numRegions_{0};
+
+  // True if stopped serving traffic. Happens if no evictions are
+  // possible due to everything being pinned. Clears when pins
+  // decrease and space can be cleared.
+  bool suspended_{false};
```

Comment:
Would we need to expose this state somehow? E.g. I don't see a public method that would tell whether the file is "suspended" or not. How would we test this state?

Reply:
Looks like this comment got lost. Repeating. "Would we need to expose this state somehow? E.g. I don't see a public method that would tell whether the file is "suspended" or not."